objectstorage/object_storage.go

```diff
@@ -401,108 +407,152 @@ func (objectStorage *ObjectStorage) Shutdown() {
 
 func (objectStorage *ObjectStorage) accessCache(key []byte, createMissingCachedObject bool) (cachedObject *CachedObjectImpl, cacheHit bool) {
 	objectStorage.flushMutex.RLock()
+	defer objectStorage.flushMutex.RUnlock()
 
 	copiedKey := make([]byte, len(key))
 	copy(copiedKey, key)
 
-	keyPartitions := objectStorage.options.keyPartitions
-	if keyPartitions == nil {
-		keyPartitions = []int{len(key)}
+	if objectStorage.options.keyPartitions == nil {
+		return objectStorage.accessNonPartitionedCache(copiedKey, createMissingCachedObject)
 	}
-	keyPartitionCount := len(keyPartitions)
 
+	return objectStorage.accessPartitionedCache(copiedKey, createMissingCachedObject)
+}
+
+func (objectStorage *ObjectStorage) accessNonPartitionedCache(key []byte, createMissingCachedObject bool) (cachedObject *CachedObjectImpl, cacheHit bool) {
+	objectKey := typeutils.BytesToString(key)
 	currentMap := objectStorage.cachedObjects
-	keyOffset := 0
 
 	objectStorage.cacheMutex.RLock()
-	var writeLocked bool
-	for i, keyPartitionLength := range keyPartitions {
-		partitionStringKey := typeutils.BytesToString(copiedKey[keyOffset : keyOffset+keyPartitionLength])
-		keyOffset += keyPartitionLength
+	alreadyCachedObject, cachedObjectExists := currentMap[objectKey]
+	if cachedObjectExists {
+		alreadyCachedObject.(*CachedObjectImpl).Retain()
+		cacheHit = true
+		cachedObject = alreadyCachedObject.(*CachedObjectImpl)
+		objectStorage.cacheMutex.RUnlock()
+		return
+	}
+	objectStorage.cacheMutex.RUnlock()
 
-		if i == keyPartitionCount-1 {
-			alreadyCachedObject, cachedObjectExists := currentMap[partitionStringKey]
-			if cachedObjectExists {
-				alreadyCachedObject.(*CachedObjectImpl).Retain()
+	if !createMissingCachedObject {
+		return
+	}
 
-				cacheHit = true
-			} else {
-				if !createMissingCachedObject {
-					objectStorage.cacheMutex.RUnlock()
+	// create a new cached object to hold the object
+	objectStorage.cacheMutex.Lock()
+	defer objectStorage.cacheMutex.Unlock()
 
-					return
-				}
+	newlyCachedObject := newCachedObject(objectStorage, key)
```

Comment:
You removed the logic with `alreadyCachedObject`, but between `cacheMutex.RUnlock` and `cacheMutex.Lock`, there could be another objected added from a different thread.